### PR TITLE
libcephfs: add ceph_fsetattr&&ceph_lchmod&&ceph_lutime

### PR DESCRIPTION
**Pull request by renhwztetecs**
export ceph_fsetattr() && ceph_lchmod() && ceph_lutime()

Signed-off-by: huanwen ren <ren.huanwen@zte.com.cn>

### CODE REVIEW

**Comment by jcsp** [comment]
Can you comment on why lchmod is needed?  It doesn't seem to part of e.g. Linux's file I/O implementation.

**Comment by jcsp** [comment]
+ Same question about lutime.

**Comment by renhwztetecs** [comment]
@jcsp 
my primitive idea，we can use  lchmod &&lutime of  user mode，not kernel mode.
:-)

**Comment by jcsp** [comment]
Right, but ordinary applications presumably don't use calls like this, since they're not part of the usual Linux stdlib interface.  So why does libcephfs need them?

**Comment by renhwztetecs** [comment]
ignorant and ill-informed, and I get it
thank you

**Comment by ffilz** [comment]
Hmm, would lchmod have allowed setting the mode of a symlink object rather than it's target? nfs-ganesha could actually use such a function, the NFS protocol allows setting attributes of symlink objects and freebsd at least supports lchmod so we could have clients trying it.

**Comment by renhwztetecs** [comment]
@ffilz
thanks for your advices
ping @jcsp

**Comment by jcsp** [comment]
Sorry for the delay.  I'm okay with this now as the fsetattr is at least a normal function on some UNIX variants.

@renhwztetecs please could you add a unit test for this in test/libcephfs/test.cc.

**Comment by jcsp** [comment]
This currently does not compile when merged to master, needs a rebase (change in definition of Client::fsetattr)

**Comment by renhwztetecs** [comment]
@jcsp I will add unit test for fsetattr

**Comment by renhwztetecs** [comment]
@jcsp
I added the unit test for fsetattr, could you take a look?

File: src/test/libcephfs/test.cc [hunk 1]
```diff
@@ -1670,3 +1671,31 @@ TEST(LibCephFS, DirChangeAttr) {
...
+  ceph_unlink(cmount, filename);
+  int fd = ceph_open(cmount, filename, O_RDWR|O_CREAT|O_EXCL, 0666);
+  ASSERT_LT(0, fd);
+  ceph_close(cmount, fd);
+
+  struct ceph_statx stx;
+  uint64_t size = 8388608;
+  stx.stx_size = size;
+
+  ASSERT_EQ(ceph_fsetattr(cmount, fd, &stx, CEPH_SETATTR_SIZE), 0);
```
Comment (jcsp) [inline src/test/libcephfs/test.cc]: You see to be passing a ceph_statx to a function that expects a `struct stat`?

Comment (renhwztetecs) [inline src/test/libcephfs/test.cc]: thanks for your reviews
I changed it

**Comment by renhwztetecs** [comment]
@thanks for your reviews
I changed it

**Comment by renhwztetecs** [comment]
@jcsp 
thanks for your reviews
I changed it

File: src/test/libcephfs/test.cc [hunk 2]
```diff
@@ -1158,6 +1158,7 @@ TEST(LibCephFS, UseUnmounted) {
   EXPECT_EQ(-ENOTCONN, ceph_lremovexattr(cmount, "/path", "name"));
   EXPECT_EQ(-ENOTCONN, ceph_setxattr(cmount, "/path", "name", NULL, 0, 0));
   EXPECT_EQ(-ENOTCONN, ceph_lsetxattr(cmount, "/path", "name", NULL, 0, 0));
+  EXPECT_EQ(-ENOTCONN, ceph_fsetattr(cmount, 0, &st, 0);
```
Comment (jcsp) [inline src/test/libcephfs/test.cc]: Missing closing `)` here.  Please do compile this test and check that it works before updating the PR.

Comment (renhwztetecs) [inline src/test/libcephfs/test.cc]: I omitted this step, the follow-up I will pay attention to this problem,
Thank you for reminding.

**Comment by jtlayton** [comment]
Ahh, just came into this late. Yeah, I agree that lchmod/lutimes isn't really needed. We have a ceph_setattrx now, which can take an AT_SYMLINK_NOFOLLOW flag. That should allow setting any of those fields on the symlink itself.

**Comment by renhwztetecs** [comment]
@jtlayton thanks
yeah, I have been removed ceph_lchmod&&ceph_lutime

**Pull request merged**